Added PRINTER_INFO_2 support and GetPrinter function

IvanRF opened the pull request:
Issue #568

dblock commented:
Don't change the whole tabs vs. spaces thing, looks like your editor brought a bunch of extra changes here.

The new mappings need tests, please.

Update CHANGELOG.

IvanRF commented:
I just updated the changelog. Also, I reduced the changes maid by the editor (eclipse).

You can test `PRINTER_INFO_2` with:
```java
for (PRINTER_INFO_2 printerInfo : WinspoolUtil.getPrinterInfo2()) {
    System.out.println(printerInfo.pPrinterName + ": " + printerInfo.pDriverName);
}
System.out.println("---ALL---");
for (PRINTER_INFO_2 printerInfo : WinspoolUtil.getAllPrinterInfo2()) {
    System.out.println(printerInfo.pPrinterName + ": " + printerInfo.pDriverName);
}
System.out.println("---BY NAME---");
PRINTER_INFO_2 printerInfo = WinspoolUtil.getPrinterInfo2("Adobe PDF");
System.out.println(printerInfo.pPrinterName + ": " + printerInfo.pDriverName);
```
I just added a new project and create the dependency with jna project to test it. I don't know how to make an specific test inside jna.

dblock commented on CHANGES.md:
```diff
@@ -7,6 +7,7 @@ Next release
 
 Features
 --------
+* [#569](https://github.com/java-native-access/jna/pull/569): Added PRINTER_INFO_2 support and GetPrinter function [@IvanRF](https://github.com/IvanRF)
```
Please make this look like the lines below. The class names are usually included with the full name like `com.sun...`, not everybody uses Windows.

dblock commented on contrib/platform/src/com/sun/jna/platform/win32/WinspoolUtil.java:
```diff
@@ -45,6 +46,62 @@
+	private static PRINTER_INFO_2[] getPrinterInfo2(int flags) {
+		IntByReference pcbNeeded = new IntByReference();
+		IntByReference pcReturned = new IntByReference();
+		Winspool.INSTANCE.EnumPrinters(flags, null, 2, null, 0, pcbNeeded, pcReturned);
```
Failures anywhere here need to raise `Win32Exception`, check explicitly for whatever the return code is that tells you to call the function again with an actual buffer.

dblock commented on contrib/platform/src/com/sun/jna/platform/win32/WinspoolUtil.java:
```diff
@@ -45,6 +46,62 @@
+		IntByReference pcReturned = new IntByReference();
+		HANDLEByReference pHandle = new HANDLEByReference();
+
+		Winspool.INSTANCE.OpenPrinter(printerName, pHandle, null);
```
Check return of this.

dblock commented on contrib/platform/src/com/sun/jna/platform/win32/WinspoolUtil.java:
```diff
@@ -45,6 +46,62 @@
+
+		PRINTER_INFO_2 pinfo2 = new PRINTER_INFO_2(pcbNeeded.getValue());
+
+		Winspool.INSTANCE.GetPrinter(pHandle.getValue(), 2, pinfo2.getPointer(), pcbNeeded.getValue(), pcReturned);
```
Check return of this.

dblock commented:
You need actual tests for these methods. Even something basic that expects a certain return code or expects at least some data.

dblock commented:
Please also squash multiple commits, thanks.

IvanRF commented:
How do I "squash" commits?

dblock commented:
There're a few good docs on this, like http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html. A totally different approach could be http://code.dblock.org/2015/08/31/getting-out-of-your-first-git-mess.html.

IvanRF commented:
For the code in WinspoolUtil, I just use the other code as a base. Do you want me to add `if (!funciton) throw new Win32Exception` in every call?

dblock commented:
WRT exception handling, it's possible old code didn't do this right, everything that can fail should be checked and raise an exception.

IvanRF commented:
I added a check for `OpenPrinter`.

The return value for the first call to `EnumPrinters` is checked by 
```
Winspool.INSTANCE.EnumPrinters(flags, null, 2, null, 0, pcbNeeded, pcReturned);
if (pcbNeeded.getValue() <= 0)
    return new PRINTER_INFO_2[0];
```
If an exception is added there, it will always throw it. I tested it.

dblock commented on CHANGES.md:
```diff
@@ -7,6 +7,7 @@ Next release
 
 Features
 --------
+* [#569](https://github.com/java-native-access/jna/pull/569): Added `com.sun.jna.platform.win32.Winspool.PRINTER_INFO_2` support and GetPrinter function in `com.sun.jna.platform.win32.Winspool` [@IvanRF](https://github.com/IvanRF)
```
The syntax of your line and the one below is different, it's `- [@](...).`, if you don't mind fixing please.

dblock commented:
Makes sense WRT the `Enum` call, it's supposed to fail every time and return the number of things back. If you _really_ want to be diligent you should be checking that it fails, then that the last error is something like insufficient buffer or something like that, but that's fine.

This still needs tests.

IvanRF commented:
Fixed the syntax in CHANGED.md, the entry before mine was wrong and I had followed that. I corrected both.

dblock commented:
Still needs tests and a rebase.

IvanRF commented:
@dblock rebase done.
Tests made by @mlfreeman2 
I think we are finally done here!

dblock commented:
i am cool with this, merging.

IvanRF commented:
Thanks! 👍
